### DISCUSSION TITLE
Mission and Goals

### DISCUSSION THREAD

**Discussion by phiarchitect** [geometor/polynumbers#1] [General]
Greetings ~

I think it is important for every good project to have a mission, a guiding principle for its existence. 

Perhaps a good place to start is the outline from my Wildberger post:
> As you have said recently, we are on a quest for patterns. The best way to identify patterns is by collecting lots of information. So, I have a proposal.  Let's assemble a team to build a polynumber analysis engine. 

> Over the last year I have become enamored with the power of Python and Sympy.  As a CAS, Sympy appears to have all the capabilities of other dedicated platforms, but the big advantage is the strong set of tools available in Python for rendering and analysis.  

> With the power of Sympy's Polynomial and Matrices libraries, we can easily generate polynumber series, extract their attributes, build matrices, and look for patterns. It would even be possible to automate lookups to OEIS from the integer sequences discovered.  

> this link returns a JSON object from OEIS from a sequence query.

> https://oeis.org/search?fmt=json&q=0,1,1,2,3,5,8,13

> Amazing!

> There are a lot of really intelligent people that lurk here that could contribute greatly to such a project with your leadership. Federico and I have been exercising these tools in different ways, and I continue to be amazed. We could start quite simply and iterate with additional analysis algorithms as we establish a foundation. 

> There is much to say about this, but perhaps will see what you and your other patrons think before extrapolating too far. 

~ φ

**Comment by phiarchitect** [comment]
Welcome @rfede 

I think a good place for us to start is organizing much of the code you have already written into a module.

I have been building a library called GEOMETOR Explorer which I think can be a good foundation for the visualization and analysis:

https://github.com/geometor/geometor-explorer

I am working on creating some explainer videos and demonstrations soon. 

Here is a rough draft of a video I created with the tools demonstrations the construction of a geometric model, then analyzing it for golden sections:

https://youtu.be/AVRZ89q5ys8

**Reply by rfede** [reply]
Gihub doesn't allow me to upload jupyter files inside the comments. I've uploaded it here https://github.com/rfede/jupyter/blob/f2eec2f169d56d8ce23b508d49078d115d088f3b/squaredSpreadFormula.ipynb . More can be done to make these functions more efficient.

**Reply by phiarchitect** [reply]
Thanks @rfede 

You are showing me many cool ways to extract content from the formulas!

We could generate a report document with analysis of each curve.

**Reply by rfede** [reply]
It seems that those intersection points are an illusion. I used matplotlib to zoom in on the points and the lines start to drift apart, although it could be for floating precision reasons
![20SpreadsCurves](https://user-images.githubusercontent.com/45501043/175820222-f4ef8969-44e2-4bf1-9496-4bc13de6418b.jpeg)
![20SpreadsCurves 1e-8+5e-1](https://user-images.githubusercontent.com/45501043/175820229-6be2caac-cb02-4083-be33-955ef545011e.jpeg)
![20SpreadsCurves 1e-10+5e-1](https://user-images.githubusercontent.com/45501043/175820237-c05c93ed-24f8-479d-93dd-1a275cff3cf1.jpeg)
![20SpreadsCurves 1e-13+5e-1](https://user-images.githubusercontent.
![20SpreadsCurves 1e-6+7 5000000000e-1](https://user-images.githubusercontent.com/45501043/175820250-a46b22d1-4cfd-416b-8db5-ed36a09b6e6d.jpeg)
com/45501
![20SpreadsCurves 1e-5+7 5000000000e-1](https://user-images.githubusercontent.com/45501043/175820248-6f2470d4-acca-47ee-9b94-6bdc024120b0.jpeg)
043/175820247-bcaa9370-b0ba-47cd-860d-583a80f35c4d.jpeg)
.

**Reply by rfede** [reply]
By plotting the points instead of letting matplotlib doing it, the meets seems to coincide
![100_spreads_curves (2x1 scale) ](https://user-images.githubusercontent.com/45501043/175829916-d1b57471-2bff-46e1-8623-55263e601ec2.jpeg)
![100_spreads_curves (2x1 scale) center meet](https://user-images.githubusercontent.com/45501043/175829920-2d906842-1b8f-4f02-8382-ed277c3cd08f.jpeg)
![100_spreads_curves (2x1 scale) upper right meet](https://user-images.githubusercontent.com/45501043/175829923-78e926ec-80f1-4359-8d46-dba6942950f7.jpeg)
.

**Reply by phiarchitect** [reply]
Your graphics are terrific!

I think if you are not specific about the number of points when plotting, Matplotlib will give you something with enough resolution for the full view - but doesn't hold up in the zoom. 

The points definitely coincide. Explorer checks to see if a new point is already in the list before adding - preventing duplicates. If they did not coincide I would have seen more than 7 points for the intersections.

**Reply by phiarchitect** [reply]
I used 1000 points to sample these curves - when zoomed you can see the issues. I think if we are analyzing intersections, we can make sure key x values are in the sample. 
![upper-left-50](https://user-images.githubusercontent.com/34562724/175840400-8472674b-9710-4163-8cae-890d589f3632.png)

**Reply by phiarchitect** [reply]
I added you with write privileges, so you can now push directly to the repo.
I would like to work with you on laying out the overall strategy and discuss how you want to participate.
I think we can create something amazing.

**Reply by rfede** [reply]
Many thanks! As long as I have the ability (much more effort than knowledge) and the time (for work reasons) to collaborate, please let me know how I can cooperate.

(edited)

**Reply by phiarchitect** [reply]
My pleasure - I plan to learn a great deal from your insight and experience.

feel free to send me an email - illumiphi@gmail.com - we can start a thread about where we are at on our respective journeys and what we hope to achieve. I suspect we have a lot of common interests.

**Reply by rfede** [reply]
I'm embarrassed for the typo in my message. I meant "As long as I have ...". Now edited.

**Reply by phiarchitect** [reply]
No worries - I caught your gist :)

**Comment by phiarchitect** [comment]
Just uploaded a new video:

https://youtu.be/5SBz-9vJ-98

**Reply by rfede** [reply]
Amazing!
I tend to believe those ghost lines are reflections and rotations of those standard spread curves.
Here is a rough example: https://www.geogebra.org/m/hbvhd7eu 
(just flash the blue g_n curves on/off and watch the visual effect)
Regarding the computational time, it shouldn't be taking more than 5 minutes for 2048 points. Which eval algorithm are you using?

**Reply by phiarchitect** [reply]
Wow - you are quite adept with Geogebra! What is your background? Have you studied with Wildberger?

I decided to use the same lamdify function (without numpy) I had previously. To be honest, I didn't under how your PolyEval worked :) I'll take another look.

**Reply by phiarchitect** [reply]
BTW - what do you mean by the blue g_n curves?

**Reply by rfede** [reply]
> 

I've uploaded a new file with explaining comments in the fork: https://github.com/rfede/polynumbers/blob/ce9a0e0d6cf0e748f9270383ee791ddb81e2c8ae/100SpreadCurves2048Points.ipynb

**Reply by rfede** [reply]
> BTW - what do you mean by the blue g_n curves?
They should appear in the left panel scrolling down

![SpreadGhostLines](https://user-images.githubusercontent.com/45501043/176426868-ec97ea15-f91f-4a52-90d8-361ffeaeca0e.jpg)

**Reply by phiarchitect** [reply]
Oh - I see. Will have to study that!

**Reply by rfede** [reply]
Another visualization example with RR (1/4 right rotations), HR (half rotations) and LR (1/4 left rotations)
https://www.geogebra.org/classic/hrdkpvyb

**Reply by phiarchitect** [reply]
So, the ghost curves can be derived from manipulations of the curves in the series?

Can you break these procedures down for me?

**Reply by rfede** [reply]
> 
Half rotation (180º) of $S_n(s)$: $x=s, y=1-S_n(s)$
```python
def getHalfRotatedPoints(n, density): 
    p = SpreadPoly(n) 
    i = Fraction(1, density) 
    k = Fraction(-1,density) 
    x, y = [], []
    for j in range(density+1):
        k += i 
        x.append(k)
        y.append(1-PolyEval(p, k))
    return x, y
```

Right rotation (90º left) of $S_n(s)$: $x = S_n(s), y=s$
```python
def getRightRotatedPoints(n, density): 
    p = SpreadPoly(n) 
    i = Fraction(1, density) 
    k = Fraction(-1,density) 
    x, y = [], []
    for j in range(density+1):
        k += i 
        x.append(PolyEval(p, k))
        y.append(k)
    return x, y

```

Left rotation (90º right) of $S_n(s)$: $x = 1 - S_n(s), y=s$
```python
def getLeftRotatedPoints(n, density): 
    p = SpreadPoly(n) 
    i = Fraction(1, density) 
    k = Fraction(-1,density) 
    x, y = [], []
    for j in range(density+1):
        k += i 
        x.append(1-PolyEval(p, k))
        y.append(k)
    return x, y
```

**Reply by phiarchitect** [reply]
Aha - so you are manipulating the x, y values. 

how do you manipulate the expression to create a new expression that produces these points?

**Reply by rfede** [reply]
In spread.py file:
for right rotation
```python
ax.plot(ysf, xsf) # interchange x and y
```
for half rotation
```python
 ysf = [1 - float(y_val) for y_val in ys] # Subtract S_n(s) from 1 
```
for left rotation:
```python
ysf = [1 - float(y_val) for y_val in ys] # Subtract S_n(s) from 1 
ax.plot(ysf, xsf) # interchange x and y
```
If you do this for all the curves, it would be difficult to distinguish them. I would select the first ones.

Another nice ghost curve is $[x, y]=[S_2(s), S_3(y)]$
![fg4](https://user-images.githubusercontent.com/45501043/176733631-306411a1-f1a5-4362-8aba-c4fae852c1db.jpeg)

**Reply by phiarchitect** [reply]
I am heading out for the day - but so glad I caught this before.

I am amazed - I will definitely incorporate this. We can render each concept in different ways - first separately, then combined. 

Thank you for your insight. 

Where are you based? My "outpost at the edge of the universe" in on the Oregon Coast.

**Reply by rfede** [reply]
I'm in Buenos Aires, Argentina.

**Reply by phiarchitect** [reply]
How exciting - people all over the world studying polynomials!

**Comment by phiarchitect** [comment]
Wildberger wants to promote the project on his main channel!! 

I would like to provide some suggestions to him for next steps. What are your thoughts?

Have you heard about the 3Blue1Brown summer contest?

**Reply by rfede** [reply]
A main topic of his thoughts is based on avoiding taking roots to find zeros or intersections. If this task cannot be achieved algebraically, although being very close to each other, the lines/curves don't meet. This leads to S.Goh factorization formula, introduced by Professor Wildberger in his videos. By factoring the difference between two curves, the intersections reveal themselves. Remarkably, the difference between two curves seems to be equal to the algebraic square root of the product of two other curves:

https://github.com/rfede/jupyter/blob/89214cc75aa2056fecea2b633116bf57b5f828bb/SpreadCurvesMeets.ipynb

**Reply by phiarchitect** [reply]
In GEOMETOR Explorer, as I add new lines and circles, I check the new element with each previous element for intersections. When points are found, I add the elements to a parents list on the point object, then add the point to a global points list. I always check for duplicates and if found, I merge the parents list with the previous point. I also add the new intersection points to the element objects. So points know all there possible parents and elements know all the points on them 

So, I want to do the same with the curves (In fact, I want to make curves a first class citizen in Explorer :). In my first attempt, I started with simply subtracting two expressions and using the Sympy roots. But Sympy fell apart with this method and was unable to handle higher degree expressions. 

I just watched the Wildberger video on the Goh Factorizations. I think I understand it better now. But I am sure you have gathered I am not a mathematician! Just a creative guy chasing what fascinates me.

So, would you like to create a function that takes two sympy polynomials and returns a list of sympy points for the intersections? I'll take care of the dedupe and ancestral lineage.

**Reply by phiarchitect** [reply]
Did you see yesterday's video:

https://youtu.be/SYjr0VtS89A

**Reply by rfede** [reply]
> Did you see yesterday's video:
> 
> https://youtu.be/SYjr0VtS89A

I saw it. It's amazing! Are you saving the interactive files for later use?

**Reply by rfede** [reply]
> In GEOMETOR Explorer, as I add new lines and circles, I check the new element with each previous element for intersections. When points are found, I add the elements to a parents list on the point object, then add the point to a global points list. I always check for duplicates and if found, I merge the parents list with the previous point. I also add the new intersection points to the element objects. So points know all there possible parents and elements know all the points on them
> 
> So, I want to do the same with the curves (In fact, I want to make curves a first class citizen in Explorer :). In my first attempt, I started with simply subtracting two expressions and using the Sympy roots. But Sympy fell apart with this method and was unable to handle higher degree expressions.
> 
> I just watched the Wildberger video on the Goh Factorizations. I think I understand it better now. But I am sure you have gathered I am not a mathematician! Just a creative guy chasing what fascinates me.
> 
> So, would you like to create a function that takes two sympy polynomials and returns a list of sympy points for the intersections? I'll take care of the dedupe and ancestral lineage.

I think I can.
ps: I'm not a mathematician either, but a lawyer :)

**Reply by phiarchitect** [reply]
> Are you saving the interactive files for later use?

I call a rendered set a session. Right now sessions are just PNG files - but could be SVG's. I have created a log file for my geometric constructions which documents each step.  I want to incorporate that for this project soon - ultimately leading to a well formatted report.

I would also like to figure out the best way to save the model and analysis for a session since they are time-consuming to create. This would allow us to restore and maybe even combine models. 

Lawyer! Never would have guessed. But I suspect you are a Renaissance Man at heart.

**Reply by rfede** [reply]
The module pickle allows to save the object as it is for later use.
To save the object, before the line "plt.show()":
```python
import pickle
pickle.dump(fig, open('FILENAME.fig.pickle', 'wb')) 
```
To open
```python
import pickle
ax = pickle.load(open(r'FILENAME.fig.pickle', 'rb'))
```

**Reply by phiarchitect** [reply]
I thought pickle might be the answer - but I had no idea you could serialize the matplotlib figure!

**Reply by rfede** [reply]
> I can
So as not to keep you waiting while I try to design something else, this sympy solve function should work fine, albeit a little slow. It should return two lists of solutions: one with rational values and another with decimal approximations. 
n and m are the integers that identify the polynomials S_n and S_m
```Python
def Meets(n, m):
    if n < m:
        n, m = m, n
    p = Spread(n + m)*Spread(n - m)
    rational, approximate = [], []
    for solution in solve(p, Symbol('s')):
        if solution.is_rational: rational.append(solution)
        else: approximate.append(N(solution))
    return rational, approximate
```

**Reply by phiarchitect** [reply]
Very interesting - thank you!

I will spend some time integrating this tonight - we'll see what we get!

But I am surprised you are multiplying the polys before solving. Shouldn't one subtract?

**Comment by phiarchitect** [comment]
For my next task - I want to animate stretching the x axis on the graph then panning across. 

What do you think it should sound like? :)

**Comment by phiarchitect** [comment]
Posted 3 videos in the last 24 hours.

Here is a playlist for all the polynumber videos

https://youtube.com/playlist?list=PLruFHGFsvC_zw19jj_n8WokICzYjSqRiM

**Comment by phiarchitect** [comment]
Behold ~
![sread_detail](https://user-images.githubusercontent.com/34562724/176980924-d827b85f-34bd-49b6-9c43-f33053aa24c0.png)
